lib/Target/Btor/BtorToBtorIRTranslation.cpp

```diff
@@ -0,0 +1,781 @@
+        case BTOR2_TAG_state:
+        case BTOR2_TAG_fair:
+        case BTOR2_TAG_justice:
+        case BTOR2_TAG_output:
+        case BTOR2_TAG_write:
+        default:
+            break;
+    }
+    return res;
+}
+
+void cleanupCreatedLines() {
+    for( auto it = reached_lines.begin(); it != reached_lines.end(); ++it ) {
+        if( it->first < 0 ) {
+            Btor2Line *line = it->second;
+            delete [] line->args;
+            delete line;
+        }
+    }
+}
+
+void createNegateLine( int64_t curAt, Btor2Line * line ) {
+    Btor2Line *res = new Btor2Line;
+    int64_t *args = new int64_t[1];
+    args[0] = curAt * -1;
+
+    res->id = curAt;
+    res->lineno = curAt;
+    res->name = "not";
+    res->tag = BTOR2_TAG_not;
+    res->sort = line->sort;
+    res->init = line->init;
+    res->next = line->next;
+    res->constant = line->constant;
+    res->symbol = line->symbol;
+    res->nargs = 1;
+    res->args = args; 
+
+    reached_lines[ curAt ] = res;
+}
+
+bool isValidChild( int64_t line ) {
+    auto tag = reached_lines.at( line )->tag;
+    if( tag == BTOR2_TAG_init || tag == BTOR2_TAG_constraint
+    ||  tag == BTOR2_TAG_next || tag == BTOR2_TAG_read
+    ||  tag == BTOR2_TAG_state || tag == BTOR2_TAG_input
+    ||  tag == BTOR2_TAG_sort || tag == BTOR2_TAG_fair 
+    || tag == BTOR2_TAG_justice || tag == BTOR2_TAG_output
+    || tag ==  BTOR2_TAG_write) {
+        return false;
+    }
+    return true;
+}
+
+void toOp( Btor2Line * line, 
+           OpBuilder builder, 
+           std::map<int64_t, Value> &cache) {
+
+    if( cache.find( line->id) != cache.end() )
+        return;
+
+    Operation * res = nullptr; 
+    std::vector<Btor2Line *> todo;
+    todo.push_back( line );
+    while( !todo.empty() ) {
+        auto cur = todo.back();
+        uint32_t oldsize = todo.size();
+        for( uint32_t i = 0; i < cur->nargs; ++i ) {
+            if( cur->args[i] > 0 && !isValidChild( cur->args[i] ) ) {
+                continue;
+            }
+
+            if( cache.find( cur->args[i] ) == cache.end() ) {
+                if( cur->args[i] < 0 ) {
+                    createNegateLine( cur->args[i], 
+                        reached_lines.at( cur->args[i] * -1 ));
+                    todo.push_back( reached_lines.at( cur->args[i] ) );
+                    todo.push_back( reached_lines.at( cur->args[i] * -1 ) );
+                } else {
+                    todo.push_back( reached_lines.at( cur->args[i] ) );
+                }
+            }
+        } 
+        if( todo.size() != oldsize ) {
+            continue;
+        }
+        if( !isValidChild( cur->id ) ) {
+            todo.pop_back();
+            continue;
+        }
+        // some operations could have been resolved prior to this
+        if( cache.find( cur-> id ) == cache.end() ) {
+            res = createMLIR( cur, builder, cache, cur->args );
+            assert( res );
+            cache[ cur->id ] = res->getResult(0);
```

Comment:
why not `cache[cur->id] = res;`?  What is `getResult(0)`?

Reply:
Discussed "irl"